File: 2016/18/18.go
package main

import (
	"fmt"

	"github.com/pjsoftware/advent-of-code/2016/lib/advent"
)

type TrapDetector struct {
	FirstRow  string
	NumRows   int
	SafeCount int
}

const TrapTile string = "^"
const TrapSym string = "❌"
const SafeTile string = "."
const SafeSym string = "🟩"

func NewDetector(row string, rows int) *TrapDetector {
	d := &TrapDetector{FirstRow: row, NumRows: rows, SafeCount: 0}
	return d
}

func main() {
	// Tests

	fmt.Print("Starting Tests:\n\n")

	advent.Test("room1", 6, Solve(NewDetector("..^^.", 3)))
	advent.Test("room2", 38, Solve(NewDetector(".^^.^.^^^^", 10)))
	advent.BailOnFail()
	fmt.Print("All tests passed!\n\n")

	// Solution

	input := advent.InputString("18")
	fmt.Printf("Solution: %d\n", Solve(NewDetector(input, 40)))
}

// Solution code

func Solve(room *TrapDetector) int {
	return room.SafeCount
}

File: 2017/18/18.rb
require_relative "../lib/advent/read"
require_relative "../lib/advent/test"

require_relative "assembly"

puts "Starting Tests:"
tests = Test.new

test_input = Read::strings("input_test.txt")
test_assem = Assembly.new
tests.test("test data solution", 4, test_assem.run(test_input))
tests.bail_on_fail

puts "All tests passed!\n---"

input = Read::strings("input_data.txt")
assem = Assembly.new
puts "Solution: #{assem.run(input)}"
